File: go-eicio/event.go
package eicio // import "github.com/decibelcooper/eicio/go-eicio"

import (
	"bytes"
	"errors"
	"fmt"
	"reflect"
	"strings"

	"github.com/golang/protobuf/proto"

	"github.com/decibelcooper/eicio/go-eicio/model"
)

// Struct representing an event either created with NewEvent() or retrieved
// with (*Reader) Next()
type Event struct {
	Header  *model.EventHeader
	payload []byte

	collCache   map[string]Collection
	namesCached []string
}

// Returns a new event with minimal initialization
func NewEvent() *Event {
	return &Event{
		Header:    &model.EventHeader{},
		collCache: make(map[string]Collection),
	}
}

// Get a list of collection names in the event
func (evt *Event) GetNames() []string {
	names := make([]string, 0)

	for _, collHdr := range evt.Header.PayloadCollections {
		names = append(names, collHdr.Name)
	}
	for name, _ := range evt.collCache {
		names = append(names, name)
	}

	return names
}

// Get a string representing the collection type
func GetType(coll Collection) string {
	return strings.TrimPrefix(proto.MessageName(coll), "eicio.model.")
}

var (
	ErrDupCollection = errors.New("duplicate collection name")
	ErrDupID         = errors.New("duplicate collection id")
)

// Add a collection to the event.  This allows automatic referencing with
// (*Event) Reference(), and queues the collection for serialization once
// (*Writer) Push() is called.  The collection may be modified further after
// adding.
func (evt *Event) Add(coll Collection, name string) error {
	for key, coll_ := range evt.collCache {
		if key == name {
			return ErrDupCollection
		}
		if coll_.GetId() != 0 && coll_.GetId() == coll.GetId() {
			return ErrDupID
		}
	}

	for _, collHdr := range evt.Header.PayloadCollections {
		if collHdr.Name == name {
			return ErrDupCollection
		}
		if collHdr.Id != 0 && collHdr.Id == coll.GetId() {
			return ErrDupID
		}
	}

	evt.collCache[name] = coll
	evt.namesCached = append(evt.namesCached, name)
	return nil
}

// Remove a collection from the event by name
func (evt *Event) Remove(name string) {
	for key, _ := range evt.collCache {
		if key == name {
			delete(evt.collCache, key)
			return
		}
	}
	for _, collHdr := range evt.Header.PayloadCollections {
		if collHdr.Name == name {
			evt.getFromPayload(name, false)
			return
		}
	}
}

// Gets a collection from the event.  The collection is deserialized upon the
// first time calling this function.  Once deserialized, the collection is
// removed from Header.PayloadCollection, and placed back into a queue for
// reserialization.  The event may be safely modified before reserializing.
func (evt *Event) Get(name string) Collection {
	if msg := evt.collCache[name]; msg != nil {
		return msg
	}

	return evt.getFromPayload(name, true)
}

// Get a unique ID for referencing collections or entries.  This is typically
// not needed, and instead (*Event) Reference() should be called.
func (evt *Event) GetUniqueID() uint32 {
	evt.Header.NUniqueIDs++
	return evt.Header.NUniqueIDs
}

// Reference a message that exists in the event.  The message may be a
// collection or entry within a collection.  In both cases, the collection must
// have been added to the event, or it must be a collection retrieved with
// (*Event) Get().
func (evt *Event) Reference(msg Message) *model.Reference {
	for _, coll := range evt.collCache {
		if coll == msg {
			collID := coll.GetId()
			if collID == 0 {
				collID = evt.GetUniqueID()
				coll.SetId(collID)
			}
			return &model.Reference{
				CollID:  collID,
				EntryID: 0,
			}
		}

		for i := uint32(0); i < coll.GetNEntries(); i++ {
			entry := coll.GetEntry(i).(Message)
			if entry == msg {
				collID := coll.GetId()
				if collID == 0 {
					collID = evt.GetUniqueID()
					coll.SetId(collID)
				}
				entryID := entry.GetId()
				if entryID == 0 {
					entryID = evt.GetUniqueID()
					entry.SetId(entryID)
				}
				return &model.Reference{
					CollID:  collID,
					EntryID: entryID,
				}
			}
		}
	}

	return nil
}

// Dereference a message from the event.  This returns a message (either
// collection or collection entry) referred to by a Reference.  The message
// must exist in the event.
func (evt *Event) Dereference(ref *model.Reference) Message {
	var refColl Collection
	for _, coll := range evt.collCache {
		if coll.GetId() == ref.CollID {
			if ref.EntryID == 0 {
				return coll
			}
			refColl = coll
			break
		}
	}
	if refColl == nil {
		for _, collHdr := range evt.Header.PayloadCollections {
			if collHdr.Id == ref.CollID {
				if refColl = evt.Get(collHdr.Name); refColl == nil {
					return nil
				}
				break
			}
		}
	}
	if refColl == nil {
		return nil
	}

	for i := uint32(0); i < refColl.GetNEntries(); i++ {
		entry := refColl.GetEntry(i).(Message)
		if entry.GetId() == ref.EntryID {
			return entry
		}
	}
	return nil
}

func (evt *Event) String() string {
	buffer := &bytes.Buffer{}

	stringBuf := fmt.Sprint(evt.Header, "\n")
	stringBuf = strings.Replace(stringBuf, " payloadCollections:", "\n    payloadCollections:", -1)
	stringBuf = strings.Replace(stringBuf, " >", ">", -1)
	fmt.Fprint(buffer, stringBuf, "\n")

	for _, name := range evt.GetNames() {
		coll := evt.Get(name)
		if coll != nil {
			fmt.Fprint(buffer, "    name:", name, " type:", GetType(coll), "\n")

			stringBuf = fmt.Sprint("        ", coll, "\n")
			stringBuf = strings.Replace(stringBuf, " entries:", "\n        entries:", -1)
			stringBuf = strings.Replace(stringBuf, " >", ">", -1)
			fmt.Fprint(buffer, stringBuf)
		}
	}

	return string(buffer.Bytes())
}

func (evt *Event) getFromPayload(name string, unmarshal bool) Collection {
	offset := uint32(0)
	size := uint32(0)
	collType := ""
	collIndex := 0
	var collHdr *model.EventHeader_CollectionHeader
	for collIndex, collHdr = range evt.Header.PayloadCollections {
		if collHdr.Name == name {
			collType = collHdr.Type
			size = collHdr.PayloadSize
			break
		}
		offset += collHdr.PayloadSize
	}
	if collType == "" {
		return nil
	}

	var coll Collection
	if unmarshal {
		msgType := proto.MessageType("eicio.model." + collType).Elem()
		coll = reflect.New(msgType).Interface().(Collection)
		if err := coll.Unmarshal(evt.payload[offset : offset+size]); err != nil {
			return nil
		}

		evt.collCache[name] = coll
	}

	evt.namesCached = append(evt.namesCached, name)
	evt.Header.PayloadCollections = append(evt.Header.PayloadCollections[:collIndex], evt.Header.PayloadCollections[collIndex+1:]...)
	evt.payload = append(evt.payload[:offset], evt.payload[offset+size:]...)

	return coll
}
func (evt *Event) flushCollCache() error {
	for _, name := range evt.namesCached {
		coll := evt.collCache[name]
		if err := evt.collToPayload(coll, name); err != nil {
			return err
		}
		delete(evt.collCache, name)
	}
	evt.namesCached = nil
	return nil
}

func (evt *Event) collToPayload(coll Collection, name string) error {
	collHdr := &model.EventHeader_CollectionHeader{}
	collHdr.Name = name
	collHdr.Id = coll.GetId()
	collHdr.Type = GetType(coll)

	collBuf, err := coll.Marshal()
	if err != nil {
		return err
	}
	collHdr.PayloadSize = uint32(len(collBuf))

	if evt.Header == nil {
		evt.Header = &model.EventHeader{}
	}
	evt.Header.PayloadCollections = append(evt.Header.PayloadCollections, collHdr)
	evt.payload = append(evt.payload, collBuf...)

	return nil
}

func (evt *Event) getPayload() []byte {
	return evt.payload
}

func (evt *Event) setPayload(payload []byte) {
	evt.payload = payload
}

type Message interface {
	proto.Message

	Marshal() ([]byte, error)
	Unmarshal([]byte) error

	GetId() uint32
	SetId(uint32)
}

type Collection interface {
	Message

	GetNEntries() uint32
	GetEntry(uint32) proto.Message
}

File: cpp-eicio/src/eicio/event.cc
#include <iostream>

#include "event.h"

using namespace google::protobuf;

using namespace eicio;

Event::Event() { header = new model::EventHeader(); }

Event::~Event() {
    if (header) delete header;
    for (auto collEntry = collCache.begin(); collEntry != collCache.end(); collEntry++) {
        delete collEntry->second;
    }
}

bool Event::Add(Message *coll, std::string name) {
    const Descriptor *desc = coll->GetDescriptor();
    const Reflection *refl = coll->GetReflection();
    const FieldDescriptor *idFieldDesc = desc->FindFieldByName("id");
    if (!idFieldDesc) return false;
    uint32 collID = refl->GetUInt32(*coll, idFieldDesc);

    for (auto collPair : collCache) {
        if (collPair.first.compare(name) == 0) return false;

        auto collEx = collPair.second;

        desc = collEx->GetDescriptor();
        refl = collEx->GetReflection();
        idFieldDesc = desc->FindFieldByName("id");
        if (!idFieldDesc) continue;
        uint32 collIDEx = refl->GetUInt32(*collEx, idFieldDesc);

        if (collIDEx == collID) return false;
    }

    for (auto collHdr : header->payloadcollections()) {
        if (collHdr.name().compare(name) == 0) return false;
        if (collHdr.id() == collID) return false;
    }

    collCache[name] = coll;
    namesCached.push_back(name);
    return true;
}

void Event::Remove(std::string collName) {
    if (collCache.find(collName) != collCache.end()) {
        delete collCache[collName];
        collCache.erase(collName);
        return;
    }

    for (auto collHdr : header->payloadcollections()) {
        if (collHdr.name().compare(collName) == 0) {
            delete getFromPayload(collName);
            return;
        }
    }
}

Message *Event::Get(std::string collName) {
    Message *msg;
    if ((msg = collCache[collName]) != NULL) return msg;

    return getFromPayload(collName);
}

std::vector<std::string> Event::GetNames() {
    std::vector<std::string> names;

    for (int i = 0; i < header->payloadcollections_size(); i++) {
        auto collHdr = header->payloadcollections(i);
        names.push_back(collHdr.name());
    }
    for (auto collEntry = collCache.begin(); collEntry != collCache.end(); collEntry++) {
        names.push_back(collEntry->first);
    }

    return names;
}

void Event::Reference(Message *msg, model::Reference *ref) {
    for (auto collPair : collCache) {
        auto coll = collPair.second;

        const Descriptor *desc = coll->GetDescriptor();
        const Reflection *refl = coll->GetReflection();

        const FieldDescriptor *idFieldDesc = desc->FindFieldByName("id");
        if (!idFieldDesc) continue;
        uint32 collID = refl->GetUInt32(*coll, idFieldDesc);

        if (coll == msg) {
            if (collID == 0) {
                collID = GetUniqueID();
                refl->SetUInt32(coll, idFieldDesc, collID);
            }

            ref->set_collid(collID);
            return;
        }

        const FieldDescriptor *entriesFieldDesc = desc->FindFieldByName("entries");
        if (!entriesFieldDesc) continue;
        RepeatedPtrField<Message> *entries = refl->MutableRepeatedPtrField<Message>(coll, entriesFieldDesc);

        for (int i = 0; i < entries->size(); i++) {
            if (&(*entries)[i] == msg) {
                const Descriptor *entryDesc = (*entries)[i].GetDescriptor();
                const Reflection *entryRefl = (*entries)[i].GetReflection();

                const FieldDescriptor *entryIDFieldDesc = entryDesc->FindFieldByName("id");
                if (!entryIDFieldDesc) continue;
                unsigned int entryID = entryRefl->GetUInt32((*entries)[i], entryIDFieldDesc);

                if (collID == 0) {
                    collID = GetUniqueID();
                    refl->SetUInt32(coll, idFieldDesc, collID);
                }
                if (entryID == 0) {
                    entryID = GetUniqueID();
                    entryRefl->SetUInt32(&(*entries)[i], entryIDFieldDesc, entryID);
                }

                ref->set_collid(collID);
                ref->set_entryid(entryID);
                return;
            }
        }
    }

    return;
}

Message *Event::Dereference(const model::Reference &ref) {
    Message *refColl = NULL;
    for (auto collPair : collCache) {
        auto coll = collPair.second;

        const Descriptor *desc = coll->GetDescriptor();
        const Reflection *refl = coll->GetReflection();

        const FieldDescriptor *idFieldDesc = desc->FindFieldByName("id");
        if (!idFieldDesc) continue;
        uint32 collID = refl->GetUInt32(*coll, idFieldDesc);

        if (collID == ref.collid()) {
            refColl = coll;
            if (ref.entryid() == 0) return refColl;
            break;
        }
    }
    if (refColl == NULL) {
        for (auto collHdr : header->payloadcollections()) {
            if (collHdr.id() == ref.collid()) {
                refColl = Get(collHdr.name());
                if (ref.entryid() == 0) return refColl;
                break;
            }
        }
    }
    if (refColl == NULL) return NULL;

    const Descriptor *desc = refColl->GetDescriptor();
    const Reflection *refl = refColl->GetReflection();

    const FieldDescriptor *entriesFieldDesc = desc->FindFieldByName("entries");
    if (!entriesFieldDesc) return NULL;
    RepeatedPtrField<Message> *entries = refl->MutableRepeatedPtrField<Message>(refColl, entriesFieldDesc);

    for (int i = 0; i < entries->size(); i++) {
        const Descriptor *entryDesc = (*entries)[i].GetDescriptor();
        const Reflection *entryRefl = (*entries)[i].GetReflection();

        const FieldDescriptor *entryIDFieldDesc = entryDesc->FindFieldByName("id");
        if (!entryIDFieldDesc) continue;
        unsigned int entryID = entryRefl->GetUInt32((*entries)[i], entryIDFieldDesc);

        if (entryID == ref.entryid()) {
            return &(*entries)[i];
        }
    }

    return NULL;
}

unsigned int Event::GetUniqueID() {
    header->set_nuniqueids(header->nuniqueids() + 1);
    return header->nuniqueids();
}

void Event::SetHeader(model::EventHeader *newHeader) {
    if (header) delete header;
    header = newHeader;
}

model::EventHeader *Event::GetHeader() { return header; }

unsigned int Event::GetPayloadSize() { return payload.size(); }

void *Event::SetPayloadSize(uint32 size) {
    payload.resize(size);
    return &payload[0];
}

unsigned char *Event::GetPayload() { return &payload[0]; }

std::string Event::GetType(Message *coll) {
    static const std::string prefix = "eicio.model.";
    return coll->GetTypeName().substr(prefix.length());
}

void Event::FlushCollCache() {
    for (int i = 0; i < namesCached.size(); i++) {
        auto name = namesCached[i];
        auto coll = collCache[name];
        collToPayload(coll, name);
        collCache.erase(name);
        namesCached.erase(namesCached.begin() + i);
        delete coll;
    }
}

Message *Event::getFromPayload(std::string name, bool parse) {
    uint32 offset = 0;
    uint32 size = 0;
    std::string collType = "";
    int collIndex = 0;
    for (int i = 0; i < header->payloadcollections_size(); i++) {
        auto collHdr = header->payloadcollections(i);
        if (name.compare(collHdr.name()) == 0) {
            collType = collHdr.type();
            size = collHdr.payloadsize();
            break;
        }
        offset += collHdr.payloadsize();
    }
    if (collType.length() == 0) {
        return NULL;
    }

    Message *coll;
    if (parse) {
        auto desc = DescriptorPool::generated_pool()->FindMessageTypeByName("eicio.model." + collType);
        if (desc == NULL) {
            return NULL;
        }
        coll = MessageFactory::generated_factory()->GetPrototype(desc)->New();

        if (!coll->ParseFromArray(&payload[0] + offset, size)) {
            delete coll;
            return NULL;
        }

        collCache[name] = coll;
        namesCached.push_back(name);
    }

    header->mutable_payloadcollections()->DeleteSubrange(collIndex, 1);
    payload.erase(payload.begin() + offset, payload.begin() + offset + size);

    return coll;
}

void Event::collToPayload(Message *coll, std::string name) {
    const Descriptor *desc = coll->GetDescriptor();
    const Reflection *ref = coll->GetReflection();

    const FieldDescriptor *idFieldDesc = desc->FindFieldByName("id");
    if (!idFieldDesc) return;

    model::EventHeader_CollectionHeader *collHdr = header->add_payloadcollections();
    collHdr->set_name(name);
    collHdr->set_id(ref->GetUInt32(*coll, idFieldDesc));
    collHdr->set_type(GetType(coll));

    const size_t byteSize = coll->ByteSizeLong();
    collHdr->set_payloadsize(byteSize);
    size_t offset = payload.size();
    payload.resize(offset + byteSize);
    unsigned char *buf = &payload[0] + offset;
    coll->SerializeToArray(buf, byteSize);
}
